lib/spack/spack/binary_distribution.py

```diff
@@ -795,33 +797,39 @@ def sign_specfile(key, force, specfile_path):
     spack.util.gpg.sign(key, specfile_path, signed_specfile_path, clearsign=True)
 
 
-def _fetch_spec_from_mirror(spec_url):
-    s = None
-    tty.debug("fetching {0}".format(spec_url))
-    _, _, spec_file = web_util.read_from_url(spec_url)
-    spec_file_contents = codecs.getreader("utf-8")(spec_file).read()
-    # Need full spec.json name or this gets confused with index.json.
-    if spec_url.endswith(".json.sig"):
-        specfile_json = Spec.extract_json_from_clearsig(spec_file_contents)
-        s = Spec.from_dict(specfile_json)
-    elif spec_url.endswith(".json"):
-        s = Spec.from_json(spec_file_contents)
-    elif spec_url.endswith(".yaml"):
-        s = Spec.from_yaml(spec_file_contents)
-    return s
-
-
-def _read_specs_and_push_index(file_list, cache_prefix, db, db_root_dir):
-    for file_path in file_list:
-        try:
-            s = _fetch_spec_from_mirror(url_util.join(cache_prefix, file_path))
-        except (URLError, web_util.SpackWebError) as url_err:
-            tty.error("Error reading specfile: {0}".format(file_path))
-            tty.error(url_err)
+def _read_specs_and_push_index(file_list, read_method, cache_prefix, db, db_root_dir, concurrency):
+    fetched_specs = []
+
+    def _fetch_spec_from_mirror(spec_url):
+        spec_file_contents = None
+        s = None
+
+        tty.debug("reading {0}".format(spec_url))
+        spec_file_contents = read_method(spec_url)
+
+        if spec_file_contents:
+            # Need full spec.json name or this gets confused with index.json.
+            if spec_url.endswith(".json.sig"):
+                specfile_json = Spec.extract_json_from_clearsig(spec_file_contents)
+                s = Spec.from_dict(specfile_json)
+            elif spec_url.endswith(".json"):
+                s = Spec.from_json(spec_file_contents)
+            elif spec_url.endswith(".yaml"):
+                s = Spec.from_yaml(spec_file_contents)
 
         if s:
-            db.add(s, None)
-            db.mark(s, "in_buildcache", True)
+            fetched_specs.append(s)
+
+    tp = multiprocessing.pool.ThreadPool(processes=concurrency)
+    try:
+        tp.map(llnl.util.lang.star(_fetch_spec_from_mirror), [(f,) for f in file_list])
```

Comment:
I think this call to map has either a race condition or is serialized by the GIL:
1.  `_fetch_spec_from_mirror` returns `None`
2. Therefore the result (we discard) from `tp.map` is a list of `None`s
3. The computations we do next are based on a global object (`fetched_specs`) that is bound to `_fetch_spec_from_mirror` as a closure

It's probably better to have `_fetch_spec_from_mirror` return a spec and turn this line into:
```suggestion
        fetched_specs = tp.map(llnl.util.lang.star(_fetch_spec_from_mirror), [(f,) for f in file_list])
```

Reply:
@scottwittenburg :point_up: This is the most important comment

Reply:
Yes, thanks, I'm attempting to address this now. :+1: